### PR TITLE
Path: helix op should not use G0 below safe height (except for retract).

### LINKED ISSUE #8140: CAM: Helix produces dangerous gcode
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Forums discussion

https://forum.freecadweb.org/viewtopic.php?f=15&t=74977

### Version

0.21 (Development)

### Full version info

```shell
OS: Fedora Linux 37 (KDE Plasma) (KDE/plasmax11)
Word size of FreeCAD: 64-bit
Version: 0.21.0.31508 (Git)
Build type: Release
Branch: master
Hash: ed8e45ac9b9196ec42572b6746b27cdd6183c903
Python 3.11.1, Qt 5.15.7, Coin 4.0.0, Vtk 9.1.0, OCC 7.6.3
Locale: English/United States (en_US)
Installed mods: 
  * lattice2 1.0.0
  * kicadStepUpMod 10.16.5
  * fasteners 0.4.51
  * freecad.gears 1.0.0
```

### Subproject(s) affected?

Path

### Issue description

The helix operation can produce G0 moves below the safeheight that are potentially dangerous.

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by itain**
fixes #8140

Forums discussion
https://forum.freecadweb.org/viewtopic.php?f=15&t=74977

- [X]  Your Pull Request meets the requirements outlined in section 5 of [CONTRIBUTING.md](https://github.com/FreeCAD/FreeCAD/blob/master/CONTRIBUTING.md) for a Valid PR

Please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

---

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [741227399](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/741227399) was triggered at [b733ce6](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/b733ce6aff882d795ebb2109846cc919e2dff72f). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [741315137](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/741315137) was triggered at [9bf2532](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/9bf253255db9bb4b873602b2c55096a7619ab956). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [741316253](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/741316253) was triggered at [ef5eb68](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/ef5eb68797409823449ddfb448a3ecdfaeeacf3c). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [746025874](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/746025874) was triggered at [0fc57a5](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/0fc57a5307845c5b5f2600cabdfdc31b78933189). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [746612489](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/746612489) was triggered at [ec840d9](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/ec840d9ded1aac1453c4a5a124cf935e21bf0fe1). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [746648792](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/746648792) was triggered at [0cab4f8](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/0cab4f832f19bc5bb09113bff06469d2cfcd7143). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [747137828](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/747137828) was triggered at [9705154](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/9705154af08f54f9d1f2104841e56e39ca516625). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [747403879](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/747403879) was triggered at [04c5c39](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/04c5c391d798296144e7d0389676f394636bf8ab). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [747650030](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/747650030) was triggered at [fe8979e](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/fe8979e1ecda78aef3ed338fb216a1082fbf3f8b). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [747709440](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/747709440) was triggered at [ce7878a](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/ce7878a1fc788b3289bc9da7619c18c56c1f06c2). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [747815440](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/747815440) was triggered at [088d903](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/088d90386d497f18198a9447c7fea693f40a4ca7). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [749106742](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/749106742) was triggered at [44616c8](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/44616c86256e8ec2492c367ac048301d6d7b2b5b). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/Path/Path/Base/Generator/helix.py [hunk 1]
```diff
@@ -229,16 +238,18 @@ def retract():
         # safe place which does not touch the inner or outer wall on all radii except
         # the first.  This is left as a future improvement.
 
-        retractcommands.append(Path.Command("G0", {"Z": startPoint.z}))
+        retractcommands.append(Path.Command("G0", {"Z": clearanceHeight if isLast else safeHeight}))
```
Comment (mlampert) [inline src/Mod/Path/Path/Base/Generator/helix.py:241]: FYI: IIRC the final retract to `ClearanceHeight` is added by the framework automatically after each operation.

File: src/Mod/Path/Path/Op/Helix.py [hunk 2]
```diff
@@ -138,7 +138,7 @@ def initCircularHoleOperation(self, obj):
             "Helix Drill",
             QT_TRANSLATE_NOOP(
                 "App::Property",
-                "Extra value to stay away from final profile- good for roughing toolpath",
+                "Extra value to stay away from final profile- good for roughing toolpath"
```
Comment (mlampert) [inline src/Mod/Path/Path/Op/Helix.py:141]: These comma cleanups are a bit suspicious. I presume you manually removed them? The formatting has been given to `black` so you might want to run that over this file.

File: src/Mod/Path/Path/Op/Helix.py [hunk 3]
```diff
@@ -198,26 +200,6 @@ def circularHoleExecute(self, obj, holes):
...
-                        "Z": obj.ClearanceHeight.Value,
-                    },
-                )
-            )
-            self.commandlist.append(
-                Path.Command(
-                    "G0", {"X": startPoint.x, "Y": startPoint.y, "Z": startPoint.z}
-                )
-            )
-
```
Comment (mlampert) [inline src/Mod/Path/Path/Op/Helix.py:220]: Isn't the move to the `startPoint` still required?

File: src/Mod/Path/Path/Base/Generator/helix.py (hunk 1 above)
Comment (itain) [reply src/Mod/Path/Path/Base/Generator/helix.py:241]: The operation may consist of **multiple** holes. I chose to retract to `ClearanceHeight` after each hole, before the G0 to the next hole.
Indeed there is an extra retract after the last hole. That does not cause any problem.

File: src/Mod/Path/Path/Op/Helix.py (hunk 2 above)
Comment (itain) [reply src/Mod/Path/Path/Op/Helix.py:141]: Probably I have removed it. I  will put it back.

Trailing commas in python argument lists are optional.
In FreeCAD most uses of QT_TRANSLATE_NOOP do not use a trailing comma. Few do.

File: src/Mod/Path/Path/Op/Helix.py (hunk 3 above)
Comment (itain) [reply src/Mod/Path/Path/Op/Helix.py:220]: This was replaced with new code in .../Path/Base/Generator/helix.py .

`startPoint` is just the hole center and not there the operation should really start.

File: src/Mod/Path/Path/Op/Helix.py (hunk 2 above)
Comment (itain) [reply src/Mod/Path/Path/Op/Helix.py]: Added commit 892933c971f33ec682fa6b7a3c22dab5216453c2 "Undo removal of comma in argument list of QT_TRANSLATE_NOOP."

**Review by mlampert — APPROVED** [review]

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [761010433](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/761010433) was triggered at [7903a4e](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/7903a4ebf221085e0208b461429e573ed43b5a5e). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by luzpaz** [comment]
Just a quick question? Is there anything left here to do ?

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [768056416](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/768056416) was triggered at [9255751](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/9255751c422b2fb3a8abba8e5271411cb56f4506). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by itain** [comment]
@luzpaz  I do not know if there is anything I can do or need to do.
I made contributions in the past that were much faster. This PR is waiting for too long.

@mlampert @sliptonic Please help!

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [772383459](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/772383459) was triggered at [df1309d](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/df1309dc870a1155669858c80c62a97531bc4a8e). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/Path/Path/Op/Helix.py (hunk 2 above)
Comment (luzpaz) [reply src/Mod/Path/Path/Op/Helix.py]: @mlampert anything else?

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [773124174](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/773124174) was triggered at [8fb0170](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/8fb0170fced80a4d790d9ff7619f57c3ad5f3402). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by sliptonic** [comment]
First, I apologize for the PR taking so long.  Between launching Ondsel and the Fosdem meetup, I've been severely slammed.

In general, the code looks good to me.  I'm a bit confused why you're passing in the safeheight and clearance height.  The generator shouldn't have any need for them.  

The safeheight is mainly used for moving between targets within an operation.  But a generator only works on a single target at a time.  

The clearance height is mainly used to move between operations when a clamp may interfere.  Again, since the generator is only working on a single target in the context of a single operation, it shouldn't have any need.

If I were building the helix generator based on my earlier assumptions,  The helix would start at the startpoint.z and end at the endpoint.z.  The generator would return the tool to the startpoint before exiting.   From there, the operation would be responsible for getting the tool into position before and after calling the generator.

File: src/Mod/Path/Path/Op/Helix.py (hunk 2 above)
Comment (mlampert) [reply src/Mod/Path/Path/Op/Helix.py]: @luzpaz I gave my approval a while ago.

**Comment by itain** [comment]
Thanks @sliptonic.
The operation can have multiple circular holes. Each hole may take multiple helices (the result is actually a circular pocket. (It was designed to make ring or annulus shaped pocked but the front-end hides this option). As you wrote the generator handles a single hole.

With my changes the redundant first move to the center of the hole is canceled. The opration moves directly to the start XY location. Then it's doing G0 to safe-height and G1 to start-height (where it can barely touch the work-piece). The retract is to the safe-height except the last helix of a hole that retract to clearance-height.
Moves between holes (of the same operation) are at clearance-height, as it was before my changes.

The clearance-height parameter is used mainly to eliminate two consequent G0 retracts to safe-height and immediately to clearance height.

* At least for LinuxCNC, probably for other controllers we **must** retract to safe-height before any horizontal move. This is because the controller adds a radius between linear moves. Safe-height must consider this radius.

**Comment by sliptonic** [comment]
> With my changes the redundant first move to the center of the hole is canceled. The opration moves directly to the start XY location. 

Where were the redundant moves coming from.  Was the generator generating more than it should or was the operation inserting them?

The operation should handle all the moves to get the cutter to the start point.  Then it should call the generator for the helix.  The generator should return all the moves to create the helix and should return the cutter back to the start point.

>Then it's doing G0 to safe-height and G1 to start-height (where it can barely touch the work-piece). The retract is to the safe-height except the last helix of a hole that retract to clearance-height. Moves between holes (of the same operation) are at clearance-height, as it was before my changes.
> 

What is the 'it' that is doing these moves?  The operation or the generator?

> The clearance-height parameter is used mainly to eliminate two consequent G0 retracts to safe-height and immediately to clearance height.
>
I still don't understand why the generator would be doing this rather than the operation.
 
>     * At least for LinuxCNC, probably for other controllers we **must** retract to safe-height before any horizontal move. This is because the controller adds a radius between linear moves. Safe-height must consider this radius.

It makes sense but not for the generator to do it.  Only for the operation.

**Comment by sliptonic** [comment]
Remember that we're trying to keep the helix generator as abstract as possible.  It might be generating helices for the operation now but it could also generate helical entry moves for a dressup or something else later.  So the interface to the generator needs to be as clean and abstract as possible.  It's the job of the operation to understand the context of the helix and do the right thing before and after.

**Comment by itain** [comment]
Current situation is that the generator does not know its boundaries and the op does not know what the generator is trying to do. The result is a possible collision. 
My changes are basically making the generator do almost all the movements, that's why it needs to know the heights.
My next suggested change (its in the forum discussion but I wanted to separate the PR) is forcing just the outer-most helix. It's useful to cut through circular holes (similar to profile hole but uses helix).
Another possible improvement is limiting the vertical speed - if the radius is small and the step is large the helix does not consider the TC max vertical speed. I do no know if in these cases we need to reduce the over-all speed, reduce the step or possibly add a new parameter to limit the helix angle.

I wonder about one point. I made some fixes around 2016-2017, mostly for the path profile op. Then the op did not really care about retract direction. If your machine supported it you could retract in X direction or event in -Z direction. I considered it nice feature even if I never used it. I think now most ops can retract only in +Z direction. Am I right? When was this feature lost?

**Comment by itain** [comment]
I got issue #8149 open and I already got the code ready to PR.

Now about this PR, just tell me if it can be merged as is or you want me to make changes.

**Comment by sliptonic** [comment]
Please remove the dependency on safe height and clearance height from the generator.  Any consideration of those things needs to be handled upstream in the operation.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147">pipeline status</a> for feature branch [PR_8147](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8147). Pipeline [778400196](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/778400196) was triggered at [41a8001](https://github.com/FreeCAD/FreeCAD/pull/8147/commits/41a80018231680096ff05c998f03ea74092c2e4c). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by itain** [comment]
The generator does need to do retractions above Start-Height. It needs these parameters!
Since Helix implements a circular pocket I checked how other pockets are implemented. Found that these parameters are passed from the Op to libarea (Path/Op/Area.py):
247         pathParams["resume_height"] = obj.SafeHeight.Value
248         pathParams["retraction"] = obj.ClearanceHeight.Value
And, off course Libarea makes use of these parameters (actually always retracting to clearance-height which IMHO is too much).

Attached example of test project using "Grid" which generates lots of retracts.
pocket

Does it make a difference if the generator is written in C++ or in Python?

**Comment by sliptonic** [comment]
> The generator does need to do retractions above Start-Height. It needs these parameters! 

This is the central question and I think we're talking past each other.  Would it be feasible to chat via telegram or in the gitter channel?  Real time chat might make it easier to understand each other.   I don't understand why the generator would need to do this and I can think of several reasons why doing it in the generator is a bad idea.  I'm not trying to be difficult, I'm just trying to understand.

> Since Helix implements a circular pocket I checked how other pockets are implemented. Found that these parameters are passed from the Op to libarea (Path/Op/Area.py): 247 pathParams["resume_height"] = obj.SafeHeight.Value 248 pathParams["retraction"] = obj.ClearanceHeight.Value And, off course Libarea makes use of these parameters (actually always retracting to clearance-height which IMHO is too much).

These aren't really relevant.  These are the older style operations that we're trying to refactor away.  They contain all the co-mixing of concepts that are causing the problems.  They shouldn't be used as a basis for future development.

> Attached example of test project using "Grid" which generates lots of retracts. pocket

I'll take a closer look this afternoon.

> Does it make a difference if the generator is written in C++ or in Python?

No, it shouldn't.  So far, all the generators are in python.

**Comment by sliptonic** [comment]
@itain Do you want to discuss synchronously?  Chat/zoom?

**Comment by petterreinholdtsen** [comment]
Is there any hope to get this change merged into master before 0.21?

**Comment by sliptonic** [comment]
> Is there any hope to get this change merged into master before 0.21?

OP hasn't responded in some time.  I'm not sure if they're interested in continuing.  I'm still happy to discuss if someone wants to adopt the work and continue it.

**Comment by sliptonic** [comment]
Closing due to lack of interest

**Pull request closed without merging**